### PR TITLE
Arf predict proba

### LINKED ISSUE #62: Should `predict_proba` be implemented for AdaptiveRandomForest?
I would like to use a AdaptiveRandomForest model and predict class probabilities with it.
To this end, I made a `predict_proba` implementation for it. Is this something that is desired to be added to the main repository?
If so, I will add documentation/tests and create a PR.

### PR DESCRIPTION
**Pull request by PGijsbers**
## Changes
Fixes #62, #63, #65.

Changes proposed in this pull request:
- Implementation of `predict_proba` for Adaptive Random Forest (#62).
- Fix a bug where weights were incorrectly passed (#63).
- Above bugfix solves #65 (technically only tested j=2,5,10 and rest still running).

### Remarks
In the discussion of #63 it was mentioned that the fix reduces performance.
However this was based on just the single unit test.
Using the same unit test, but changing the test set (by increasing the initial value of cnt), we see the following values:

| cnt  | master  | PR  |
|---|---|---|
|  0 | 0.6531  | 0.4694  |
|  1 | 0.5306 | 0.4694  |
|  2 | 0.4082  | 0.4694  |
|  3 | 0.5310  | 0.5306  |
|  4 | 0.3877  | 0.5510  |
|  5 | 0.5306  | 5714  |
|  6 | 0.5306  | 0.5918  |
|  7 | 0.4287  | 0.4898  |
|  8 | 0.408  | 0.5510  |
|  9 | 0.5510  | 0.5918  |

While we can do more thorough evaluations, for me it seems that the initial set-up just was an outlier, and there is no real performance decrease.
Further motivation is that if I increase the training set size on the PR version, performance does increase.
On the PR branch, using 150 samples for training results in a performance of 0.4694 (with cnt=0 determining the test set). Increasing this to 1500 increases performance to 0.5714, increasing the training set again to 15000 results in a performance of 0.7142.

## Checklist

- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] Tests are included for new functionality.
- [x] Travis CI build passes.
- [x] Test Coverage is maintained (threshold 0.2%)

@scikit-multiflow/scikit-multiflow-repo-admins

### CODE REVIEW

**Comment by PGijsbers** [comment]
There seems to be a difference between the predictions produced on the CI server and on my local machine. My first guess is that this is somehow because of the weights now being assigned correctly (as Poisson RV). However predictions are consistent locally across tests.  
Second option would be that some changes to the master branch might have changed the predictions. I will merge the up-to-date master into my development branch to check.

**Comment by PGijsbers** [comment]
Merging the development branch into an up-to-date master branch locally also passes all tests.
As far as I can tell, the random state is controlled just fine (variable set in initialization and used when generating a weight). Any ideas?

**Comment by jacobmontiel** [comment]
Hello @PGijsbers , some comments:

- Could you explain what do you mean by `cnt`?
- In `adaptive_random_forests.py` you import `check_is_fitted` but can't find where is being used?

**Comment by PGijsbers** [comment]
With `cnt` I mean the local variable in the test function `test_adaptive_random_forests`. Changing that was just a quick and dirty way to test with different test sets.

I noticed that too in the PR, made the change locally but did not push yet because I was trying to figure out why the CI test generates different predictions.

**Comment by PGijsbers** [comment]
Pushed the commit which removes the unused import. CI results should not change.

**Comment by jacobmontiel** [comment]
Regarding the failed tests, the failure on different python versions is quite strange and unexpected. A similar behaviour was observed recently 21cb5de303b0f0f5bf71ef055d92f43534a95975. 

I will run a local test of your branch to confirm.

**Comment by PGijsbers** [comment]
Good point. I do run Python 3.6 locally which does actually pass the test on Travis.

**Comment by jacobmontiel** [comment]
Tests for ARF fail locally in Python 3.5.

```
...
>       assert np.alltrue(predictions == last_version_predictions)
E       assert False
E        +  where False = <function alltrue at 0x7f6f73baf7b8>([1, 1, 0, 1, 1, 0, ...] == [1, 1, 1, 1, 1, 0, ...]
E        +    where <function alltrue at 0x7f6f73baf7b8> = np.alltrue
...
tests/meta/test_adaptive_random_forests.py:42: AssertionError
...
>       assert np.alltrue(predictions == last_version_predictions)
E       assert False
E        +  where False = <function alltrue at 0x7f6f73baf7b8>(array([1, 1, ..., 1, 0, 0, 0]) == [1, 1, 0, 1, 1, 0, ...]
E        +    where <function alltrue at 0x7f6f73baf7b8> = np.alltrue
...
tests/meta/test_adaptive_random_forests.py:83: AssertionError
...
```


Also got failures on evaluators unit tests, did you merge the master branch? There are changes missing but

**Comment by PGijsbers** [comment]
No I did not merge the master branch (as that should be done automatically), but I'll push a merge commit.
It seems that you encounter the same issue as Travis.

**Comment by PGijsbers** [comment]
Merged latest version of master into the development branch. No merge conflicts, all tests pass locally with Python 3.6.

**Comment by jacobmontiel** [comment]
Ingore my last comment about the evaluators, it was a local issue when I did the pull.

**Comment by jacobmontiel** [comment]
I can confirm that tests work locally on Python 3.6.5.

I will also take a look into this. before making the decision.

**Comment by PGijsbers** [comment]
I found a bug in my implementation which caused `predict_proba` to fail when called to predict for `X` with more than one sample.

The cause was the implementation which normalized the probabilities to 1. I now use scikit-learn's `preprocessing.normalize` instead (normalizing with l1 norm). This works both in the single and multi sample setting.

I added a unit test which calls `predict_proba` in this multi sample setting. It did fail with the old version, but passes the fix.

All tests pass locally (on Python 3.6).

**Comment by jacobmontiel** [comment]
#### Regarding failing tests.
- The problem is unrelated to Travis, I can confirm that locally the same test fails on Python 3.4.5 and 3.5.5 while passing on 3.6.5 and vice-versa.
- The root cause is still unkown but seems (directly) unrelated to the code changes in the pull request.
- My suggestion is to create the test that passes on Travis with Python 3.4.x and 3.5.x and **temporary** disable the assert line for 3.6.5:
```python
import sys

if sys.version[2] != 6:
    assert ...
```

A new issue will be created before merging the PR for further investigation and to enable the test(s) for all Python versions.

Note: If you are using anaconda, then you can create a virtual environment with a different Python version to run the tests.

----
#### Other comments
During review I found a bug in the code unrelated to this PR/issue:

This section
https://github.com/scikit-multiflow/scikit-multiflow/blob/6ae5e9fad1f2b22b1735c355aacffdca20fa1fa5/src/skmultiflow/meta/adaptive_random_forests.py#L491-L498

should be inside of this `if` block:
https://github.com/scikit-multiflow/scikit-multiflow/blob/6ae5e9fad1f2b22b1735c355aacffdca20fa1fa5/src/skmultiflow/meta/adaptive_random_forests.py#L469

Could you add this fix to your PR?

Let me know if can be of any help.

**Comment by PGijsbers** [comment]
Done

**Comment by jacobmontiel** [comment]
Thanks @PGijsbers !

Can you disable the tests for the versions where they fail in order to get a pass from Travis? I will use the commit for the temporary deactivation in a new issue.

**Comment by jacobmontiel** [comment]
Thanks @PGijsbers for your work, patience and your understanding on this change.

I have created #68 to further investigate the issue with the test.

This change is ready to merge with the master.
